### PR TITLE
Improve perf of hash one-shots on Win10

### PR DESCRIPTION
**Pull request by GrabYourPitchforks**
Uses bcrypt pseudo-handles (see https://docs.microsoft.com/windows/win32/seccng/cng-algorithm-pseudo-handles) with the `BCryptHash` routine to get a modest perf improvement in the one-shot hash algortihms when running on Windows 10. For downlevel Windows platforms, continues to use the existing "catch `EntryPointNotFound` and fall back to `BCryptHashData`" logic.

The Win10-specific routine also has a safety net such that if any of the invariants are violated (unknown algorithm or destination too short), it falls back to the downlevel code paths for additional error handling.

No change on other OSes.

|   Method | Toolchain | HashAlgorithm | SourceLengthInBytes |     Mean |   Error |  StdDev | Ratio |
|--------- |---------- |-------------- |-------------------- |---------:|--------:|--------:|------:|
| HashData |    master |          SHA1 |                   0 | 336.4 ns | 0.71 ns | 0.66 ns |  1.00 |
| HashData |     proto |          SHA1 |                   0 | 288.8 ns | 2.17 ns | 1.81 ns |  0.86 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |          SHA1 |                  64 | 420.8 ns | 2.22 ns | 1.85 ns |  1.00 |
| HashData |     proto |          SHA1 |                  64 | 374.6 ns | 1.35 ns | 1.27 ns |  0.89 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |          SHA1 |                 256 | 651.0 ns | 3.40 ns | 3.02 ns |  1.00 |
| HashData |     proto |          SHA1 |                 256 | 609.0 ns | 4.11 ns | 3.85 ns |  0.94 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |        SHA256 |                   0 | 274.7 ns | 1.33 ns | 1.24 ns |  1.00 |
| HashData |     proto |        SHA256 |                   0 | 226.2 ns | 1.13 ns | 0.94 ns |  0.82 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |        SHA256 |                  64 | 302.8 ns | 1.53 ns | 1.43 ns |  1.00 |
| HashData |     proto |        SHA256 |                  64 | 255.2 ns | 0.51 ns | 0.45 ns |  0.84 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |        SHA256 |                 256 | 390.4 ns | 1.10 ns | 0.92 ns |  1.00 |
| HashData |     proto |        SHA256 |                 256 | 348.5 ns | 1.90 ns | 1.78 ns |  0.89 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |        SHA512 |                   0 | 493.0 ns | 2.43 ns | 2.27 ns |  1.00 |
| HashData |     proto |        SHA512 |                   0 | 455.6 ns | 1.32 ns | 1.10 ns |  0.92 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |        SHA512 |                  64 | 498.4 ns | 1.02 ns | 0.91 ns |  1.00 |
| HashData |     proto |        SHA512 |                  64 | 468.5 ns | 2.48 ns | 2.07 ns |  0.94 |
|          |           |               |                     |          |         |         |       |
| HashData |    master |        SHA512 |                 256 | 947.0 ns | 4.13 ns | 3.86 ns |  1.00 |
| HashData |     proto |        SHA512 |                 256 | 907.9 ns | 4.28 ns | 4.01 ns |  0.96 |

### CODE REVIEW

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs [hunk 1]
```diff
@@ -58,33 +69,77 @@ public static unsafe int HashData(string hashAlgorithmId, ReadOnlySpan<byte> sou
...
+                }
 
-                        return hashSize;
+                NTSTATUS ntStatus;
+                fixed (byte* pSrc = &MemoryMarshal.GetReference(source))
+                fixed (byte* pDest = &MemoryMarshal.GetReference(destination))
+                {
+                    try
+                    {
+                        ntStatus = Interop.BCrypt.BCryptHash((uint)algHandle, null /* pbSecret */, 0 /* cbSecret */, pSrc, source.Length, pDest, digestSizeInBytes);
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs:127]: The comments could be named for parameters, too, I think:

```suggestion
                        ntStatus = Interop.BCrypt.BCryptHash((uint)algHandle, pbSecret: null, cbSecret: 0, pSrc, source.Length, pDest, digestSizeInBytes);
```

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs [hunk 2]
```diff
@@ -58,33 +69,77 @@ public static unsafe int HashData(string hashAlgorithmId, ReadOnlySpan<byte> sou
...
+
+                return hashSize;
+            }
+
+            private static unsafe bool HashDataUsingPseudoHandle(string hashAlgorithmId, ReadOnlySpan<byte> source, Span<byte> destination, out int hashSize)
+            {
+                hashSize = default;
+
+                Interop.BCrypt.BCryptAlgPseudoHandle algHandle;
+                int digestSizeInBytes;
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs:82]: I wonder if pulling this hashSize logic up and completely avoid the call to `BCryptGetProperty` would gain any worthwhile wins even in the "slow" path.

**Review by vcsjones — COMMENTED** [review]
Neat! Thank you!

**Comment by GrabYourPitchforks** [comment]
@vcsjones I addressed your second feedback item by querying the hash size right after we create the algorithm, then shoving that information into the algorithm handle cache. It means we only take the lookup hit once. For the one-shot algos, I think it's still appropriate to hard-code the expected digest length since we need to maintain the friendly-name -\> pseudo-alg handle mapping ourselves anyway.

**Comment by GrabYourPitchforks** [comment]
CI failures are tracked by https://github.com/dotnet/runtime/issues/40416 and https://github.com/dotnet/runtime/issues/40564.

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs [hunk 3]
```diff
@@ -33,58 +34,100 @@ public static class OneShotHashProvider
...
+                if (destination.Length < digestSizeInBytes)
                 {
-                    throw new CryptographicException();
+                    Debug.Fail("Caller should have checked length.");
+                    return false;
                 }
 
-                if (!s_useCompatOneShot)
+                NTSTATUS ntStatus;
+                fixed (byte* pSrc = &MemoryMarshal.GetReference(source))
```
Comment (vcsjones) [inline src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs:110]: A question: why the change to use `MemoryMarshal.GetReference`?

**Review by vcsjones — APPROVED** [review]

File: src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs (hunk 3 above)
Comment (GrabYourPitchforks) [reply src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs:110]: Habit, muscle memory, whatever you want to call it. :)

Comment (gfoidl) [reply src/libraries/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/HashProviderDispenser.Windows.cs:110]: The way via [GetPinnableReference](https://github.com/dotnet/runtime/blob/4a58c28b80022311ef6aa982a7c8975af509ef53/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L250-L256) has an extra length check, that can be avoided here, as we know the span won't be empty. So the [codegen](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBEBLbAcwDsJcGDmFwBuGvSasASgFdeQ/DBYBJBTCgQADgGUNAN2EwxNCQGYG83NgBmMJqQYAxCBAY0A3jQY+mF4ACeGPYAsgAU0jDYACYA8rwANgE6Wti8ADyBwQB8DLgQslBgMACU7tS+DF4Vlb42HAgw0QxhWTAAVAxaGFAMALx5BUWl3rU+1WNjxADsDO3dUOI1YwC+oz5ry+WVxP5B9gByEVFxicmpGW25+YXFJetVD5VsMMCy3CwAgri4GhhhN2GDAAhANeLIEgkSktJttJvVGs1WvtOgt+gwAGQhGD4aABELYKC4AAW2ASLAA4jAMJE7LBeMUAUM7vctpUJrDfDM5gsYZNNpVNisgA) is better (maybe the reason for Levi's muscle memory to kick in 😉).

**Review by bartonjs — APPROVED** [review]
Seems like a lot of work and complexity for a relatively trivial gain, but since it's written... OK.

**Comment by bartonjs** [comment]
```
    System.Net.Http.Functional.Tests.SocketsHttpHandler_ConnectionFactoryTest.CustomConnectionFactory_SyncRequest_Fails [FAIL]
      Assert.Equal() Failure
      Expected: HostNotFound
      Actual:   Unknown
      Stack Trace:
        /_/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs(165,0): at System.Net.Http.Functional.Tests.SocketsHttpHandler_ConnectionFactoryTest.CustomConnectionFactory_SyncRequest_Fails()
        /_/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs(165,0): at System.Net.Http.Functional.Tests.SocketsHttpHandler_ConnectionFactoryTest.CustomConnectionFactory_SyncRequest_Fails()
        --- End of stack trace from previous location ---
    System.Net.Http.Functional.Tests.SyncHttpHandler_HttpClientHandler_Authentication_Test.Proxy_DomainJoinedProxyServerUsesKerberos_Success [SKIP]
      Condition(s) not met: "IsDomainJoinedServerAvailable"
    System.Net.Http.Functional.Tests.SocketsHttpHandler_ConnectionFactoryTest_Http2.CustomConnectionFactory_SyncRequest_Fails [FAIL]
      Assert.Equal() Failure
      Expected: HostNotFound
      Actual:   Unknown
      Stack Trace:
        /_/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs(165,0): at System.Net.Http.Functional.Tests.SocketsHttpHandler_ConnectionFactoryTest.CustomConnectionFactory_SyncRequest_Fails()
        /_/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs(165,0): at System.Net.Http.Functional.Tests.SocketsHttpHandler_ConnectionFactoryTest.CustomConnectionFactory_SyncRequest_Fails()
        --- End of stack trace from previous location ---
```

This same test (or a very similar looking one) has failed on at least 3 reruns for the last commit of the PR.  Looks like it only happens on Win7, and I haven't seen it on other PRs (but maybe I'm unobservant).

I don't see how it could necessarily be related, but the consistency bothers me.

**Comment by vcsjones** [comment]
@bartonjs maybe the PR just needs to merge in master? https://github.com/dotnet/runtime/issues/40564

**Comment by bartonjs** [comment]
Odd, I thought that every test rerun tried stacking the PR diff on top of master freshly... but I agree with your assessment that this is a known/fixed issue.

:shrug:
